src/passes/RemoveUnusedBrs.cpp

```diff
@@ -701,7 +702,51 @@ struct RemoveUnusedBrs : public WalkerPass<PostWalker<RemoveUnusedBrs>> {
     struct Optimizer : public PostWalker<Optimizer> {
       bool worked = false;
 
+      // Set of seen labels, used to generate a fresh label.
+      std::unordered_set<Name> labels;
+      // Used to generate fresh labels.
+      Index freshNameHint = 0;
+      // Whether we need to uniquify the labels at the end.
+      bool hasLabelConflict = false;
+
+      void noteLabel(Name label) {
+        auto [it, inserted] = labels.insert(label);
+        if (!inserted) {
+          hasLabelConflict = true;
+        }
+      }
+
+      Name freshLabel() {
+        auto label = Names::getValidName(
+          "_br_on_opt",
+          [&](Name name) { return !labels.count(name); },
+          freshNameHint++);
+        labels.insert(label);
+        return label;
+      }
+
+      void visitBlock(Block* curr) { noteLabel(curr->name); }
+
+      void visitLoop(Loop* curr) { noteLabel(curr->name); }
+
+      void visitBreak(Break* curr) { noteLabel(curr->name); }
+
+      void visitTry(Try* curr) {
+        noteLabel(curr->name);
+        noteLabel(curr->delegateTarget);
+      }
+
+      void visitRethrow(Rethrow* curr) { noteLabel(curr->target); }
+
+      void visitSwitch(Switch* curr) {
+        for (auto label : curr->targets) {
+          noteLabel(label);
+        }
+        noteLabel(curr->default_);
+      }
```

Comment:
Thanks, will take a look at BranchUtils. Running UniqueNameMapper unconditionally if we ever generate a fresh name works, but in practice should almost never be necessary, so it would be nice to be able to be more precise. Even better if we could be more precise without the brittleness of the current code.

Reply:
I don't follow. Aren't we good already since we only map names if we encounter this very rare `br_on_cast` situation? Some heavier work in UniqueNameManner in that case seems fine to me.